core/trie/trie.go

```diff
@@ -709,3 +709,49 @@ func (t *Trie) dump(level int, parentP *Key) {
 		storage: t.storage,
 	}).dump(level+1, t.rootKey)
 }
+
+// Iterate the trie from startValue in ascending order until the consumer returned false or an error occur or end of
+// trie was reached. Return true if end of trie is reached.
+// TODO: its much more efficient to iterate from the txn level. But even without that, if the leaf are ordered correctly,
+// block cache should have a pretty good hit rate.
+func (t *Trie) Iterate(startValue *felt.Felt, consumer func(key, value *felt.Felt) (bool, error)) (bool, error) {
+	if startValue == nil {
+		startValue = &felt.Zero
+	}
+	startKey := t.feltToKey(startValue)
+
+	return t.doIterate(&startKey, t.rootKey, consumer)
+}
+
+func (t *Trie) doIterate(startKey, key *Key, consumer func(key, value *felt.Felt) (bool, error)) (bool, error) {
+	if key == nil {
+		return false, nil
+	}
+
+	node, err := t.storage.Get(key)
+	if err != nil {
+		return false, err
+	}
+
+	if key.Len() == t.height {
+		if startKey.CmpAligned(key) > 0 {
+			return true, nil
+		}
+		keyAsFelt := key.Felt()
+		return consumer(&keyAsFelt, node.Value)
+	}
```

Comment:
I find this logic a bit confusing. 

For example, if the trie has keys 0,1,2,3 and 5, and startKey is 4, then the consumer will be called. But the consumer in the snap_support_test.go returns true (assuming the maxNode and limitQuery parameters are not set), which suggest that there are no more nodes to the right of key 4, which is incorrect, right?